src/achievements/achievement.cpp

```diff
@@ -47,179 +47,203 @@ Achievement::~Achievement()
 {
-    out << "        <achievement id=\"" << m_id << "\" "
+    out << "        <achievement id=\"" << getID() << "\" "
         << "achieved=\"" << m_achieved << "\"";
-    if (isAchieved())
-    {
-        out << "/>\n";
-        return;
-    }
-
-    out << ">\n";
-    std::map<std::string, int>::iterator i;
-    for (i = m_progress_map.begin(); i != m_progress_map.end(); ++i)
-    {
-        out << "          <" << i->first
-            << " value=\"" << i->second << "\"/>\n";
-    }
-    out << "        </achievement>\n";
+ 
+    out << "/>\n";
 }   // save
 
 // ----------------------------------------------------------------------------
-/** Returns the value for a key.
+/** Returns how much of an achievement has been achieved in the form n/m.
+ *  The AchievementInfo adds up the number of goals if there are several,
+ *  or take the target value of the goal if there is only one.
+ *  This do the same, but with achieved value or fullfilled goals.
  */
-int Achievement::getValue(const std::string & key)
+irr::core::stringw Achievement::getProgressAsString()
 {
-    if (m_progress_map.find(key) != m_progress_map.end())
-        return m_progress_map[key];
-    return 0;
-}
+    //TODO : add a progress computation function.
+    int progress = 0;
+    irr::core::stringw target = getInfo()->toString();
+
+    // For now return N/N in case of an achieved achievement.
+    if (m_achieved)
+        return target + "/" + target;
+
+    return StringUtils::toWString(progress) + "/" + target;
+}   // getProgressAsString
+
 // ----------------------------------------------------------------------------
-/** Resets all currently key values to 0. Called if the reset-after-race flag
- *  is set for the corresponding AchievementInfo.
+/** Set any leaf of the progress goal tree whose type matches the
+ *  goal_string to the value passed as parameter.
+ *  The goal string can contain a logical prefix.
+ *  If it is LOGC- ; the update is for the current value of a
+ *  resetable counter. It is applied if the parent node is of type
+ *  SUM or AND-AT-ONCE, ignored otherwise.
+ *  If it is LOGM- ; the update is for the highest achieved value of a
+ *  resetable counter. It is appliedif the parent node is of type
+ *  AND or OR, ignored otherwise.
+ *  If there is no logical prefix, the new value is set in all cases.
+ *
+ *  If the leaf has an operator defined, this will trigger an update of the
+ *  relevant values.
+ *  If the leaf's new value match or exceed its target goal value,
+ *  a check for the achievement's completions is triggered.
  */
-void Achievement::reset()
+void Achievement::setGoalValue(std::string &goal_string, int value)
 {
-    std::map<std::string, int>::iterator iter;
-    for (iter = m_progress_map.begin(); iter != m_progress_map.end(); ++iter)
+    if(m_achieved) // This should not happen, but it costs little to double-check
+        return;
+
+    bool AO = true;
+    bool SAAO = true;
```

Comment:
What does AO and SAAO mean? I assume And-Or, Sum-And-At-Once, but clearer names would make the code easier to read